### ISSUE TITLE
OpenSSL stack exhaustion

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#7161]
I've found a possible OpenSSL stack exhaustion. In case when tls1_mac() returns 0 (in my code it happened bacause of a buggy engine), we get infinite recursion causing stack exhaustion.

Here is an example of how stack does look like:

```
3830 0x00007ffff7b626b2 in do_ssl3_write (s=s@entry=0x55555583e640, type=type@entry=21, buf=0x555555840250 "\002P", pipelens=pipelens@entry=0x7fffffff9c80, numpipes=numpipes@entry=1, 
    create_empty_fragment=create_empty_fragment@entry=0, written=0x7fffffff9c88) at ssl/record/rec_layer_s3.c:1064
3831 0x00007ffff7b6d0e9 in ssl3_dispatch_alert (s=0x55555583e640) at ssl/s3_msg.c:78
3832 0x00007ffff7b626b2 in do_ssl3_write (s=s@entry=0x55555583e640, type=type@entry=21, buf=0x555555840250 "\002P", pipelens=pipelens@entry=0x7fffffffad90, numpipes=numpipes@entry=1, 
    create_empty_fragment=create_empty_fragment@entry=0, written=0x7fffffffad98) at ssl/record/rec_layer_s3.c:1064
3833 0x00007ffff7b6d0e9 in ssl3_dispatch_alert (s=0x55555583e640) at ssl/s3_msg.c:78
3834 0x00007ffff7b626b2 in do_ssl3_write (s=s@entry=0x55555583e640, type=type@entry=21, buf=0x555555840250 "\002P", pipelens=pipelens@entry=0x7fffffffbea0, numpipes=numpipes@entry=1, 
    create_empty_fragment=create_empty_fragment@entry=0, written=0x7fffffffbea8) at ssl/record/rec_layer_s3.c:1064
3835 0x00007ffff7b6d0e9 in ssl3_dispatch_alert (s=0x55555583e640) at ssl/s3_msg.c:78
3836 0x00007ffff7b626b2 in do_ssl3_write (s=s@entry=0x55555583e640, type=type@entry=22, buf=buf@entry=0x555555841de0 "\024", pipelens=pipelens@entry=0x7fffffffd020, numpipes=<optimized out>, 
    create_empty_fragment=create_empty_fragment@entry=0, written=0x7fffffffd008) at ssl/record/rec_layer_s3.c:1064
3837 0x00007ffff7b632bd in ssl3_write_bytes (s=s@entry=0x55555583e640, type=type@entry=22, buf_=0x555555841de0, len=<optimized out>, written=written@entry=0x7fffffffd178) at ssl/record/rec_layer_s3.c:613
3838 0x00007ffff7b976e7 in ssl3_do_write (s=0x55555583e640, type=22) at ssl/statem/statem_lib.c:46
3839 0x00007ffff7b8e976 in statem_do_write (s=0x55555583e640) at ssl/statem/statem.c:706
3840 write_state_machine (s=0x55555583e640) at ssl/statem/statem.c:859
3841 state_machine (s=0x55555583e640, server=0) at ssl/statem/statem.c:440
3842 0x00007ffff7b63764 in ssl3_write_bytes (s=0x55555583e640, type=23, buf_=0x555555830830, len=0, written=0x7fffffffd498) at ssl/record/rec_layer_s3.c:383
3843 0x00007ffff7b77970 in ssl_write_internal (s=s@entry=0x55555583e640, buf=buf@entry=0x555555830830, num=num@entry=0, written=written@entry=0x7fffffffd498) at ssl/ssl_lib.c:1938
3844 0x00007ffff7b77b85 in SSL_write (s=s@entry=0x55555583e640, buf=buf@entry=0x555555830830, num=num@entry=0) at ssl/ssl_lib.c:1952
3845 0x00005555555aefd4 in s_client_main (argc=<optimized out>, argv=<optimized out>) at apps/s_client.c:2840
3846 0x000055555559c680 in do_cmd (prog=0x55555582afc0, argc=7, argv=0x7fffffffdf70) at apps/openssl.c:564
3847 0x0000555555587edc in main (argc=7, argv=0x7fffffffdf70) at apps/openssl.c:183
```

I suspect there may be some other circumstances causing such a behavior.

**Comment by mspncp** [comment]
@beldmit thanks for your report. I've taken the liberty of adding triple backquotes around your stack trace to make it more readable. Hope you don't mind.

**Comment by bernd-edlinger** [comment]
@beldmit can you reproduce this?
If yes, could you try if this fixes the problem?

```
diff --git a/ssl/statem/statem.c b/ssl/statem/statem.c
index d75f9ea..5602429 100644
--- a/ssl/statem/statem.c
+++ b/ssl/statem/statem.c
@@ -119,7 +119,8 @@ void ossl_statem_fatal(SSL *s, int al, int func, int reason, const char *file,
                        int line)
 {
     /* We shouldn't call SSLfatal() twice. Once is enough */
-    assert(s->statem.state != MSG_FLOW_ERROR);
+    if (s->statm.in_init && s->statem.state == MSG_FLOW_ERROR)
+      return;
     s->statem.in_init = 1;
     s->statem.state = MSG_FLOW_ERROR;
     ERR_put_error(ERR_LIB_SSL, func, reason, file, line);
```

**Comment by mattcaswell** [comment]
I'd be tempted to put the `ERR_put_error()` call before the return in that fix. Otherwise we wouldn't even know that an error occurred sending the alert.

**Comment by bernd-edlinger** [comment]
Yes.  I can reproduce the assertion when I inject a SSLfatal near line 1064 in ssl/record/rec_layer_s3.c

**Comment by bernd-edlinger** [comment]
Merged to 1.1.1 and master.  Thanks!

**Comment by mohitsahni** [comment]
> I'd be tempted to put the `ERR_put_error()` call before the return in that fix. Otherwise we wouldn't even know that an error occurred sending the alert.

I hit this issue if a malloc fails during do_ssl3_write call.
